File: Subprogramas/Sobrecarregados/Sobrecarregado.cpp
//C++

#include <iostream>

using namespace std;

class Sobrecarregado {
public:
    void imprimir(const string& s) {
        cout << s << std::endl;
    }

    void imprimir(int i) {
        cout << i << endl;
    }
};

int main() {
    Sobrecarregado p;
    p.imprimir("Hello");  // chama o método que aceita string
    p.imprimir(10);       // chama o método que aceita int
    return 0;
}

File: Subprogramas/Sobrecarregados/sobrecarregado.rs
//Rust não suporta diretamente sobrecarga de subprogramas. 
//A Sobrecarga é implementada por meio de traits (semelhantes às interfaces em outras linguagens).

struct Sobrecarregado;

impl Sobrecarregado {
    fn print_str(&self, s: &str) {
        println!("{}", s);
    }

    fn print_int(&self, i: i32) {
        println!("{}", i);
    }
}

fn main() {
    let p = Sobrecarregado;
    p.print_str("Hello");  // chama o método que aceita string
    p.print_int(10);       // chama o método que aceita int
}
